### ISSUE TITLE
EV_module- vehicle not going for charging.

### ISSUE THREAD

**Issue by prvnpandey** [matsim-org/matsim-code-examples#293]
I have tested the EV example and it is working fine. when I am trying to model my test scenario then vehicles are not going for charging even the SOC becomes 0. why this happening so, is something I am missing to tell the controller.

![1 individual_soc_time_profiles_Line](https://user-images.githubusercontent.com/38705566/75204547-768b8f00-5771-11ea-9ad4-5d765195ec46.png)
![1 average_soc_time_profiles_Line](https://user-images.githubusercontent.com/38705566/75204583-8d31e600-5771-11ea-9f17-6cb81083233b.png)
![1 charger_occupancy_time_profiles_Line](https://user-images.githubusercontent.com/38705566/75204627-a63a9700-5771-11ea-8982-5e1a6b35bb7c.png)
@jfbischoff  @michalmac

**Comment by tschlenther** [comment]
I guess that, at the moment, we do not have any charging behaviour logic for urban scenarios. We will do something about this, soon. That means, agents will anticipate the leg during which the SOC falls under a critical level and put a charging activity in their plan before that.
I guess that until we have done that, you will have to deal with the current state. 
Even after that, it will still be technically possible that vehicles run out of energy but keep going..

**Comment by michalmac** [comment]
Which example is that? E-taxi, E-drt or something else?

**Comment by prvnpandey** [comment]
first, I run the RunEvExample and checked the result produced by example when I am running the example provided in the folder ev/example it running perfect and the vehicle is going for charging while going from a to b. I just replaced plans and network and tried to run the scenario for testing but the vehicle is not going for charging even SOC becomes zero. @michalmac

**Comment by prvnpandey** [comment]
The vehicle is not going for charging when they have a series of legs. EV example is working fine, with only one leg why it so and how to proceed by including multiple legs of agents plan for ev simulation.

Is something I am missing please guide me in the right way.
Thanks and Regards
@michalmac @jfbischoff

**Comment by mattjweist** [comment]
@prvnpandey 
Hi Praveen, 
Were you ever able to come up with a solution to this? My output graphs look very similar. The vehicles go to the charging stations and wait for some arbitrary amount of time, but the state of charge does not go back up.
Thanks,
Matt

**Comment by prvnpandey** [comment]
@mattjweist 
Yes because in logic and for routing instead of comparing soc each time the battery cap has been compared whenever agent starts their leg, this is written like this because this library is developed for a long haul trip. If you find your way to deal with this, I think it will work for you. 

Regards
Praveen

**Comment by tschlenther** [comment]
We are currently working on an urban charging behavioral model, as mentioned above and are finally making progress (unfortunately were distracted by other things such as corona modeling...).
In a reasonable amount of time this should be available for users.

**Comment by mattjweist** [comment]
@prvnpandey did you just do something like this in EVNetworkRoutingModule? This seems to work to an extent, but it's not foolproof. If there is a large queue at the charging station, only the first (number of available plugs) vehicles get a full charge. This is evident when the initial departure time for every agent is equal. 

double capacity = ev.getBatteryCapacity() * (0.8 + random.nextDouble() * 0.18);
double initialCapacity = ev.getInitialSoc() * (0.8 + random.nextDouble() * 0.18);
double numberOfStops = Math.ceil( (estimatedOverallConsumption - initialCapacity) / capacity);
if (numberOfStops < 1) {
return basicRoute;
} else {
List<Link> stopLocations = new ArrayList<>();
double currentConsumption = 0;
for (Map.Entry<Link, Double> e : estimatedEnergyConsumption.entrySet()) {
	currentConsumption += e.getValue();
	// First stop
	if (currentConsumption > initialCapacity && stopLocations.isEmpty()) {
		stopLocations.add(e.getKey());
		currentConsumption = 0;
	}
	// all remaining stops
	else if (currentConsumption > capacity) {
		stopLocations.add(e.getKey());
		currentConsumption = 0;
	}
}
